### PR TITLE
adding one point to whitelist contract constraints migration - signed…

### PR DESCRIPTION
**Pull request by snedamle**
Adding one point to whitelist contract constraints migration - signed CorDapp JAR must be registered with the CZ network operator

### CODE REVIEW

**Comment by josecoll** [comment]
@lankydan I've noticed you changed this doc to remove the step (old step 2/4) that @snedamle is stating is required. 
Please can you review and double check whether this step is indeed still required or not (since the code changed to support automatic transitioning of CZ to Signature constraints)

**Comment by josecoll** [comment]
@snedamle you appear to have reverted back to the previous commit for this documentation.
This is not correct as the behaviour has changed (eg. code now supports auto-transitioning of CZ whitelisted states).

File: docs/source/cordapp-constraint-migration.rst [hunk 1]
```diff
@@ -165,6 +174,12 @@ The code below details how to explicitly add a Signature Constraint:
...
-3. As a node operator you need to add the new signed version of the contracts CorDapp to the ``/cordapps`` folder together with the latest version of the flows jar.
+4. As a node operator you need to add the new signed version of the contracts CorDapp to the ``/cordapps`` folder together with the latest version of the flows jar.
    Please also ensure that the original unsigned contracts CorDapp is removed from the ``/cordapps`` folder (this will already be present in the
    nodes attachments store) to ensure the lookup code in step 3 retrieves the correct signed contract CorDapp JAR.
+
+Later releases
+~~~~~~~~~~~~~~
+
+The next version of Corda will provide automatic transition of *CZ whitelisted* constrained states. This means that signed CorDapps running on a Corda 4.x node will
+automatically propagate any pre-existing on-ledger *CZ whitelisted* constrained states (and generate *signature* constrained outputs).sn
```
Comment (lankydan) [inline docs/source/cordapp-constraint-migration.rst]: There is a random "sn" at the end here

**Comment by lankydan** [comment]
@josecoll The cordapp must be whitelisted to transition a state with a whitelist constraint to a signature constraint. But that is only included in 4.3 not 4.0.

Ah I see, the change to the docs that I made should not have been made under 4.0...

**Comment by lankydan** [comment]
I think the change is fine in general. Really, I need to update this doc myself to add a 4.3 version to detail the content in "later releases". As later releases, is actually implemented in the version that the docs are for...

**Comment by josecoll** [comment]
On a related point, we now have a documentation guideline that states "not to mention forward version functionality" (so we should remove the section on "Later releases" completely). @edward-prosser can you confirm this guideline is in force.

**Comment by josecoll** [comment]
@snedamle to summarise:

Corda OS 4.0 and Corda OS 4.1 correctly detail the 4 steps required for these versions:
https://docs.corda.net/releases/release-V4.0/cordapp-constraint-migration.html
https://docs.corda.net/releases/release-V4.1/cordapp-constraint-migration.html

Corda OS 4.3 (aka. master, next version of Corda OS, there is no OS 4.2) correctly details the 3 steps for 4.3 functionality:
https://docs.corda.net/releases/head/cordapp-constraint-migration.html

**Comment by josecoll** [comment]
@lankydan 
I recommend we explicitly reference the Corda version variable (`|corda_version|`)so that the headings do not confuse the end reader.

eg. header in the RST source
```
Corda |corda_version|
~~~~~~~~~~~~~~~~~~~~~
```

**Comment by snedamle** [comment]
ok got it, adding the jar hash to whitelistzone param in network param will be automatically handled in version 4.3. so no changes required for 4.3. closing this merge request then.

**Comment by lankydan** [comment]
The cordapp must still be in the whitelistzone param. The only automatic bit is the migration to signature constraints.

File: docs/source/cordapp-constraint-migration.rst [hunk 2]
```diff
@@ -129,13 +129,22 @@ Corda 4.0 requires some additional steps to consume and evolve pre-existing on-l
...
 
-2. Any flow that builds transactions using this CorDapp will automatically transition states to use the ``SignatureAttachmentConstraint`` if
-   no other constraint is specified. Therefore, there are two ways to alter the existing code.
+2. The new Corda 4 signed CorDapp JAR must be registered with the CZ network operator (as whitelisted in the network parameters which are distributed
+   to all nodes in that CZ). The CZ network operator should check that the JAR is signed and not allow any more versions of it to be whitelisted in the future.
+   From now on the development organisation that signed the JAR is responsible for signing new versions.
 
-   * Do not specify a constraint
-   * Explicitly add a Signature Constraint
+   The process of CZ network CorDapp whitelisting depends on how the Corda network is configured:
```
Comment (lankydan) [inline docs/source/cordapp-constraint-migration.rst]: Ok, so i think I have not explained this very well. And now it is finally clicking in my head. I think we should add the docs back in that you are suggesting (adding the hash to the whitelist).

I think this point (2.) should stay as what I wrote. But with an extra clause:

"Any flow that builds transactions using this CorDapp will automatically transition states to use the ``SignatureAttachmentConstraint`` if no other constraint is specified __and the CorDapp continues to be whitelisted__. Therefore, there are two ways to alter the existing code."

Then add your addition into point 3 explaining how to whitelist a cordapp. Or add how to whitelist a cordapp as a note box under part 2.

The part 3 you added is not needed.

**Comment by lankydan** [comment]
I would also recommend changing the header from 4.0 to 4.3 as @josecoll suggested. I think at that point it should be good. Sorry for all the confusion

File: docs/source/cordapp-constraint-migration.rst [hunk 3]
```diff
@@ -63,10 +63,10 @@ Hash constraints migration
 
 .. note:: These instructions only apply to CorDapp Contract JARs (unless otherwise stated).
 
-Corda 4.0
+Corda 4.3
```
Comment (josecoll) [inline docs/source/cordapp-constraint-migration.rst]: Use variable so we don't need to change this in future releases:
change `4.3` for `|corda_version|`

File: docs/source/cordapp-constraint-migration.rst [hunk 4]
```diff
@@ -63,10 +63,10 @@ Hash constraints migration
...
 ~~~~~~~~~
 
-Corda 4.0 requires some additional steps to consume and evolve pre-existing on-ledger **hash** constrained states:
+Corda 4.3 requires some additional steps to consume and evolve pre-existing on-ledger **hash** constrained states:
```
Comment (josecoll) [inline docs/source/cordapp-constraint-migration.rst]: same: change `4.3` for `|corda_version|`

File: docs/source/cordapp-constraint-migration.rst [hunk 5]
```diff
@@ -106,21 +106,14 @@ Corda 4.0 requires some additional steps to consume and evolve pre-existing on-l
...
-
 .. _cz_whitelisted_constraint_migration:
 
 CZ whitelisted constraints migration
 -------------------------------------
 
 .. note:: These instructions only apply to CorDapp Contract JARs (unless otherwise stated).
 
-Corda 4.0
+Corda 4.3
```
Comment (josecoll) [inline docs/source/cordapp-constraint-migration.rst]: ditto: change `4.3` for `|corda_version|`

**Review by josecoll — CHANGES_REQUESTED** [review]
Looks good - just some minor edits on the version number.

**Review by josecoll — APPROVED** [review]
Looks good to me.
@lankydan please give it a final thumbs up!

**Review by lankydan — APPROVED** [review]
👍 LGTM

**Pull request merged**